File: demo/flutter/flutternew/lib/viewmodel/login_vm.dart
import 'dart:async';

import 'package:flutter/foundation.dart';

import '../data/remote/auth_remote_api_source.dart';
import '../domain/dto/req/user_login_request.dart';
import '../domain/model/user_session_model.dart';
import '../manager/app_session.dart';
import '../manager/user_manager.dart';

sealed class LoginIntent {
  const LoginIntent();
}

class LoginUpdateAccount extends LoginIntent {
  final String account;
  const LoginUpdateAccount(this.account);
}

class LoginUpdatePassword extends LoginIntent {
  final String password;
  const LoginUpdatePassword(this.password);
}

class LoginSubmit extends LoginIntent {
  const LoginSubmit();
}

class LoginGoRegister extends LoginIntent {
  const LoginGoRegister();
}

class LoginSelectSavedAccount extends LoginIntent {
  final String account;
  const LoginSelectSavedAccount(this.account);
}

class LoginTouristAccess extends LoginIntent {
  const LoginTouristAccess();
}

class LoginState {
  final String account;
  final String password;
  final bool isLoading;
  final List<UserSessionModel> savedSessions;

  const LoginState({
    required this.account,
    required this.password,
    required this.isLoading,
    required this.savedSessions,
  });

  bool get canSubmit => account.trim().isNotEmpty && password.isNotEmpty && !isLoading;

  LoginState copyWith({
    String? account,
    String? password,
    bool? isLoading,
    List<UserSessionModel>? savedSessions,
  }) {
    return LoginState(
      account: account ?? this.account,
      password: password ?? this.password,
      isLoading: isLoading ?? this.isLoading,
      savedSessions: savedSessions ?? this.savedSessions,
    );
  }
}

sealed class LoginEffect {
  const LoginEffect();
}

class LoginNavigateToMain extends LoginEffect {
  const LoginNavigateToMain();
}

class LoginNavigateToRegister extends LoginEffect {
  const LoginNavigateToRegister();
}

class LoginShowToast extends LoginEffect {
  final String message;
  const LoginShowToast(this.message);
}

class LoginVm extends ChangeNotifier {
  LoginVm({
    AuthRemoteApiSource? remoteApiSource,
    UserManager? userManager,
  }) : _remoteApiSource = remoteApiSource ?? AuthRemoteApiSource(),
       _userManager = userManager ?? UserManager.instance {
    _loadSavedAccounts();
  }

  final AuthRemoteApiSource _remoteApiSource;
  final UserManager _userManager;
  final _effectController = StreamController<LoginEffect>.broadcast();

  LoginState _state = const LoginState(
    account: '',
    password: '',
    isLoading: false,
    savedSessions: [],
  );
  LoginState get state => _state;
  Stream<LoginEffect> get effects => _effectController.stream;

  void processIntent(LoginIntent intent) {
    if (intent is LoginUpdateAccount) {
      _state = _state.copyWith(account: intent.account);
      notifyListeners();
    } else if (intent is LoginUpdatePassword) {
      _state = _state.copyWith(password: intent.password);
      notifyListeners();
    } else if (intent is LoginSelectSavedAccount) {
      _selectSavedAccount(intent.account);
    } else if (intent is LoginSubmit) {
      _submit();
    } else if (intent is LoginGoRegister) {
      _effectController.add(const LoginNavigateToRegister());
    } else if (intent is LoginTouristAccess) {
      _touristAccess();
    }
  }

  Future<void> _loadSavedAccounts() async {
    final list = (await _userManager.getAllUsers())
        .where((e) => !(e.accessToken == 'tourist' || e.account == 'tourist' || e.userId == 1))
        .toList();
    _state = _state.copyWith(savedSessions: list);
    if (list.isNotEmpty) {
      _state = _state.copyWith(
        account: _state.account.isEmpty ? list.first.account : _state.account,
        password: _state.password.isEmpty ? list.first.password : _state.password,
      );
    }
    notifyListeners();
  }

  void _selectSavedAccount(String account) {
    final hit = _state.savedSessions.where((e) => e.account == account).toList();
    if (hit.isEmpty) return;
    final selected = hit.first;
    _state = _state.copyWith(account: selected.account, password: selected.password);
    notifyListeners();
  }

  Future<void> _submit() async {
    if (!_state.canSubmit) {
      _effectController.add(const LoginShowToast('请输入账号和密码'));
      return;
    }
    _state = _state.copyWith(isLoading: true);
    notifyListeners();
    try {
      final auth = await _remoteApiSource.login(
        UserLoginRequest(
          account: _state.account.trim(),
          password: _state.password,
        ),
      );
      if ((auth.userId ?? 0) <= 0) {
        _effectController.add(const LoginShowToast('登录失败'));
      } else {
        final session = UserSessionModel(
          userId: auth.userId ?? 0,
          account: auth.account ?? '',
          name: auth.name ?? '',
          avatarUrl: auth.avatarUrl ?? '',
          accessToken: auth.accessToken ?? '',
          password: _state.password,
          isCurrent: true,
          lastLoginAt: DateTime.now().millisecondsSinceEpoch,
        );
        await _userManager.saveCurrentUser(session);
        AppSession.instance.updateUserId(session.userId);
        _effectController.add(const LoginNavigateToMain());
      }
    } catch (e) {
      _effectController.add(LoginShowToast('登录失败: $e'));
    } finally {
      _state = _state.copyWith(isLoading: false);
      notifyListeners();
      _loadSavedAccounts();
    }
  }

  Future<void> _touristAccess() async {
    final session = UserSessionModel(
      userId: 1,
      account: 'tourist',
      name: '游客',
      avatarUrl: '',
      accessToken: 'tourist',
      password: '',
      isCurrent: true,
      lastLoginAt: DateTime.now().millisecondsSinceEpoch,
    );
    await _userManager.saveCurrentUser(session);
    AppSession.instance.updateUserId(session.userId);
    _effectController.add(const LoginNavigateToMain());
  }

  @override
  void dispose() {
    _effectController.close();
    super.dispose();
  }
}

File: demo/kmp/shared/src/commonMain/kotlin/com/vectordemo/viewModel/activity/LoginVm.kt
package com.vectordemo.viewModel.activity

import androidx.lifecycle.viewModelScope
import com.vectordemo.dataSource.remote.UserRemoteApiSource
import com.vectordemo.di.AppContainer
import com.vectordemo.domain.model.user.UserSessionModel
import com.vectordemo.manager.user.UserManager
import com.vectordemo.viewModel.BaseVm
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class LoginVm(
    private val userManager: UserManager,
    private val userRemote: UserRemoteApiSource,
) : BaseVm() {
    private val _uiState = MutableStateFlow(LoginState())
    val uiState: StateFlow<LoginState> = _uiState.asStateFlow()

    private val _dataState = MutableStateFlow(LoginDataState())
    val dataState: StateFlow<LoginDataState> = _dataState.asStateFlow()

    private val _effect = Channel<LoginEffect>(Channel.BUFFERED)
    val effect: Flow<LoginEffect> = _effect.receiveAsFlow()

    init {
        loadSavedAccounts()
    }

    fun processIntent(intent: LoginIntent) {
        when (intent) {
            is LoginIntent.UpdateAccount -> _uiState.update { it.copy(account = intent.account) }
            is LoginIntent.UpdatePassword -> _uiState.update { it.copy(password = intent.password) }
            is LoginIntent.SelectSavedAccount -> selectSavedAccount(intent.account)
            LoginIntent.SubmitLogin -> submitLogin()
            LoginIntent.NavigateToRegister -> sendEffect(LoginEffect.NavigateToRegister)
            LoginIntent.TouristAccess -> touristAccess()
        }
    }

    private fun loadSavedAccounts() = viewModelScope.launch {
        val sessions = userManager.getAllUsers()
        _dataState.update { it.copy(savedUserSessions = sessions) }
    }

    private fun selectSavedAccount(account: String) {
        _dataState.value.savedUserSessions.firstOrNull { it.account == account }?.let { hit ->
            _uiState.update { it.copy(account = hit.account, password = hit.password) }
        }
    }

    private fun submitLogin() {
        val state = _uiState.value
        if (!state.canSubmit) {
            sendEffect(LoginEffect.ShowToast("请输入账号和密码"))
            return
        }

        _uiState.update { it.copy(isLoading = true) }
        viewModelScope.launch {
            try {
                val session = userRemote.login(state.account.trim(), state.password)
                userManager.saveCurrentUser(session)
                AppContainer.updateUserId(session.userId)
                AppContainer.updateToken(session.accessToken)
                _uiState.update { it.copy(isLoading = false) }
                loadSavedAccounts()
                sendEffect(LoginEffect.NavigateToMain)
            } catch (_: Throwable) {
                _uiState.update { it.copy(isLoading = false) }
                sendEffect(LoginEffect.ShowToast("登录失败"))
            }
        }
    }

    private fun touristAccess() {
        viewModelScope.launch {
            val tourist = UserSessionModel(
                userId = 1L,
                account = "tourist",
                name = "游客",
                avatarUrl = "",
                accessToken = "tourist",
                password = "",
            )
            userManager.saveCurrentUser(tourist)
            AppContainer.updateUserId(tourist.userId)
            AppContainer.updateToken(tourist.accessToken)
            loadSavedAccounts()
            sendEffect(LoginEffect.NavigateToMain)
        }
    }

    private fun sendEffect(effect: LoginEffect) = viewModelScope.launch { _effect.send(effect) }
}

sealed class LoginIntent {
    data class UpdateAccount(val account: String) : LoginIntent()
    data class UpdatePassword(val password: String) : LoginIntent()
    data class SelectSavedAccount(val account: String) : LoginIntent()
    data object SubmitLogin : LoginIntent()
    data object NavigateToRegister : LoginIntent()
    data object TouristAccess : LoginIntent()
}

data class LoginState(
    val account: String = "",
    val password: String = "",
    val isLoading: Boolean = false,
) {
    val canSubmit: Boolean get() = account.isNotBlank() && password.isNotBlank() && !isLoading
}

data class LoginDataState(
    val savedUserSessions: List<UserSessionModel> = emptyList(),
)

sealed class LoginEffect {
    data object NavigateToMain : LoginEffect()
    data object NavigateToRegister : LoginEffect()
    data class ShowToast(val message: String) : LoginEffect()
}
